### DISCUSSION TITLE
Configure Maven Private Artifactory by default using config.js on self hosted

### DISCUSSION THREAD

**Discussion by erzz** [renovatebot/renovate#9126] [Request Help]
I am really struggling to get this work having tried dozens of permutations!

**Simple requirements:**

* If the package begins `com.myorg` then use our artifactory maven repo `https://artifactory.myorg.com:443/artifactory/projectname-maven-virtual`
* Else use default maven.org etc

**We are given url / credentials etc:**

repo: `https://artifactory.myorg.com:443/artifactory/projectname-virtual`
username: `xyz`
password: `a.JWT.token` (an access token in Artifactory parlance I believe)

**Testing with PAW / Curl**

* Using curl/PAW I can get a 200 response using `Authorization: Bearer a.JWT.token` header
* Using curl/PAW I can get a 200 response using basic auth `-u username:a.JWT.token`

Configuring the following options in the config.js I get only 401's when renovate-runner tries to get the dependencies in artifactory (and then it falls back to maven.org as expected - but misses entirely our private dependencies of course)

**config.js using basic auth**
```
module.exports = {
  trustLevel: "high",
  hostRules: [
    {
      hostType: "maven",
      domainName: "artifactory.myorg.com",
      authType: "Basic",
      username: process.env.ARTIFACTORY_USER,
      password: process.env.ARTIFACTORY_JWT
    },
  ]
};
```

**config.js using token**

```
module.exports = {
  trustLevel: "high",
  hostRules: [
    {
      hostType: "maven",
      domainName: "artifactory.myorg.com",
      authType: "Bearer",
      token: process.env.ARTIFACTORY_JWT
    },
  ]
};
```

I have also experimented with and without package rules such as 

```
packageRules: [
           {
             managers: ["maven"],
             registryUrls: [
               "https://artifactory.myorg.com:443/artifactory/projectname-virtual",
             ],
           },
         ],
```

ARTIFACTORY_USER and ARTIFACTORY_JWT are CICD variables that I can see stamped out in the debug log so I know that they are accessible and expanded

e.g. 

```
DEBUG: Checking for config file in /builds/retail-pim/supporting-infrastructure/renovate-bot/module.exports = {
         hostRules: [
           {
             hostType: "maven",
             hostName: "artifactory.myorg.com",
             authType: "Bearer",
             token:
               "[MASKED]",
           },
         ],
         packageRules: [
           {
             managers: ["maven"],
             registryUrls: [
               "https://artifactory.myorg.com:443/artifactory/projectname-virtual",
             ],
           },
         ],
       };
```
And a few dozen variations :)

Can anyone help me out? Am I doing it totally wrong? Is it even possible to do it this way via config.js for all projects?

**Comment by rarkins** [comment]
> * Using curl/PAW I can get a 200 response using `Authorization: Bearer a.JWT.token` header

If this is the case then the simplest hostRules would be:

```
    {
      hostName: "artifactory.myorg.com",
      token: process.env.ARTIFACTORY_JWT
    }
```

Please note that `artifactory.myorg.com` is not a domain name (it's a subdomain/hostname), so that could be one reason of why things weren't working (there may be others). I've changed it to `hostName` in the above

**Reply by erzz** [reply]
Thanks for the answer - still no luck though I am afraid. If I take the debug logs for a single specific package

```
DEBUG: Looking up com.github.spotbugs:spotbugs-annotations in repository https://repo.maven.apache.org/maven2/ (repository=groupname/projectname)

DEBUG: Found 42 new releases for com.github.spotbugs:spotbugs-annotations in repository https://repo.maven.apache.org/maven2/ (repository=groupname/projectname)

DEBUG: Looking up com.github.spotbugs:spotbugs-annotations in repository https://artifactory.myorg.com:443/artifactory/projectname/ (repository=groupname/projectname)

DEBUG: Looking up com.github.spotbugs:spotbugs-annotations in repository https://repo.spring.io/milestone/ (repository=groupname/projectname)

DEBUG: Dependency lookup unauthorized. Please add authentication with a hostRule (repository=groupname/projectname)
       "failedUrl": "https://artifactory.myorg.com/artifactory/projectname-virtual/com/github/spotbugs/spotbugs-annotations/maven-metadata.xml"

DEBUG: Content is not found for Maven url: https://artifactory.myorg.com/artifactory/rpim-maven-virtual/com/github/spotbugs/spotbugs-annotations/maven-metadata.xml (repository=groupname/projectname)

DEBUG: Looking up com.github.spotbugs:spotbugs-annotations in repository https://repo.spring.io/milestone/ (repository=groupname/projectname)

DEBUG: Content is not found for Maven url: https://repo.spring.io/milestone/com/github/spotbugs/spotbugs-annotations/maven-metadata.xml (repository=groupname/projectname)

```

So it is attempting at least - and attempting the correct path - but still a 401

```
{
                 "datasource": "maven",
                 "depName": "com.github.spotbugs:spotbugs-annotations",
                 "currentValue": "4.2.2",
                 "fileReplacePosition": 5682,
                 "registryUrls": [
                   "https://repo.maven.apache.org/maven2",
                   "https://artifactory.myorg.com:443/artifactory/projectname-virtual",
                   "https://repo.spring.io/milestone"
                 ],
                 "depIndex": 8,
                 "updates": [],
                 "warnings": [],
                 "sourceUrl": "https://github.com/spotbugs/spotbugs",
                 "homepage": "https://spotbugs.github.io/",
                 "fixedVersion": "4.2.2"
               },
```

Current config is simply:

```
module.exports = {
  trustLevel: "high",
  hostRules: [
    {
      hostName: "artifactory.myorg.com",
      token: process.env.ARTIFACTORY_JWT
    },
  ]
};

```

**Comment by erzz** [comment]
I don't know if its possibly a reason for it - but when running in PAW for example the initial response is in fact a 302 and I need to specifically enable to follow links and Redirect Authorization

![image](https://user-images.githubusercontent.com/58994586/111172199-e8fe2580-85a5-11eb-8c19-a9c3a94d64ac.png)

Not sure if that is handled in Renovate?  If I disable the redirction of auth then I can recreate the 401

**Reply by rarkins** [reply]
Can you capture the full flow of URLs? If it is redirected to a different hostName then that could definitely be a reason why it's not authenticated.

**Reply by erzz** [reply]
Not sure if you know a better way ... but running

```
curl -L -v "https://artifactory.myorg.com/artifactory/projectname-virtual" \
     -H 'Authorization: Bearer a.JWT.token'  2>&1 | egrep "^> (Host:|GET)"

> GET /artifactory/projectname-virtual HTTP/1.1
> Host: artifactory.myorg.com
> GET /artifactory/projectname-virtual/ HTTP/1.1
> Host: artifactory.myorg.com
```

🤷🏻 

Wait I checked the curl log in more detail and see

```
 HTTP/1.1 302 Found
< Date: Mon, 15 Mar 2021 15:04:49 GMT
< Content-Length: 0
< Connection: keep-alive
< X-JFrog-Version: Artifactory/7.11.2 71102900
< X-Artifactory-Id: e4bfc857e94da9b34d5023946ae9d2f711a8373e
< X-Artifactory-Node-Id: artifactory-artifactory-ha-member-1
< Location: https://artifactory.myorg.com:443/artifactory/projectname-virtual/
```

See its a 302 from  `artifactory.myorg.com` to `artifactory.myorg.com:443`

**Reply by rarkins** [reply]
I don't know a better way. It looks like it's just redirecting to add a trailing slash?

**Reply by erzz** [reply]
I have tried adding the "port" variant as a second alternative like below, but no dice.

```
module.exports = {
  trustLevel: "high",
  hostRules: [
    {
      hostName: "artifactory.myorg.com",
      token: process.env.ARTIFACTORY_JWT
    },
    {
      hostName: "artifactory.myorg.com:443",
      token: process.env.ARTIFACTORY_JWT
    },
  ]
};
```

**Reply by erzz** [reply]
> I don't know a better way. It looks like it's just redirecting to add a trailing slash?

Sorry the example I posted first was curling specifically :443 - I updated to show happens excluding port from the initial request. Its the same thing and it seems there is a specific redirect to :443.

**Reply by rarkins** [reply]
hostnames don't have ports, so the second rule is invalid

**Reply by erzz** [reply]
It was worth a try :)

**Reply by viceice** [reply]
Did you tried `baseUrl`?

**Reply by erzz** [reply]
Just tried:

```
module.exports = {
  trustLevel: "high",
  hostRules: [
    {
      domainName: "myorg.com",
      token: process.env.ARTIFACTORY_JWT
    }
  ]
};
```
and 

```
module.exports = {
  trustLevel: "high",
  hostRules: [
    {
      baseUrl: "https://artifactory.myorg.com",
      token: process.env.ARTIFACTORY_JWT
    }
  ]
};
```

Same result in both cases

**Reply by viceice** [reply]
Looks like `got` doesn't pass the auth header after redirect?

`https://artifactory.myorg.com:443/`  looks very ugly. maybe another artifactory issue?

**Reply by erzz** [reply]
Yeah I suspect that the appending of 443 is something to do with their HA set up.

But absolutely my gut-feeling too with `got`

**Reply by viceice** [reply]
yes, maybe got does see a new host because of the port change `` -> `443`  and removed the auth header

**Reply by viceice** [reply]
Maybe we can try to build a test case with nock to reproduce the issue

**Reply by rarkins** [reply]
> yes, maybe got does see a new host because of the port change `` -> `443` and removed the auth header

Possibly when following a redirect like this:
- `got` removes the existing authorization header
- Our auth logic is not reapplied because got handles it internally

We could work around it by handling redirects ourselves in the datasource layer but it could be a lot of work.

**Reply by viceice** [reply]
I think we should start to reproduce in our test cases before doing any bigger refactoring

**Reply by viceice** [reply]
ok, added simple test case, but cant reproduce:

https://github.com/renovatebot/renovate/blob/e4e9332c447cbc9d34e9f1d18b24686c946747b3/lib/datasource/maven/index.spec.ts#L336-L392


So renovate is doing the following requests:

https://github.com/renovatebot/renovate/blob/e4e9332c447cbc9d34e9f1d18b24686c946747b3/lib/datasource/maven/__snapshots__/index.spec.ts.snap#L3-L46

@erzz Can you curl those urls to check if there is any redirect on them?

**Reply by erzz** [reply]
To save the sanitisation and possibly omitting too much - I emailed some curls directly to you @rarkins 

Summary being the 302 is only for the repo root https://artifactory.myorg.com/artifactory/projectname-virtual and everything else is a direct response

**Reply by viceice** [reply]
As I know renovate does not call the virtual repo root, only matadata.xml and pom.

**Comment by rarkins** [comment]
Quick question: which type and version of Renovate are you running? e.g. npm CLI, Docker CLI, or WhiteSource Renovate On-Premises? I want to check that you've got a very recent version in use

**Reply by erzz** [reply]
Morning!

We are using the gitlab-runner.

**Current .gitlab-ci.yml**
```
include:
  - project: 'renovate-bot/renovate-runner'
    file: '/templates/renovate.gitlab-ci.yml'

renovate:
  variables:
    RENOVATE_EXTRA_FLAGS: <list of projects>
    RENOVATE_ONBOARDING: 'true'
    RENOVATE_ONBOARDING_CONFIG: '{"$$schema": "https://docs.renovatebot.com/renovate-schema.json", "extends": ["config:base", ":semanticCommits"], "branchPrefix": "rv8/", "prConcurrentLimit": 3, "branchConcurrentLimit": 3, "hashedBranchLength": 20 }'
    RENOVATE_CONFIG_FILE: $CI_PROJECT_DIR/config.js
    LOG_LEVEL: debug
```

**Reply by viceice** [reply]
Why you don't use the dind version? it should be faster, because it doesn't have to download the big fat renovate full image. It's instead a smaller slim image and docker sidecar container for other tools, which are downloaded on demand.

**Reply by erzz** [reply]
As we run our own gitlab runners in our own GKE cluster, we avoid DinD and privileged pods!

**Reply by erzz** [reply]
But Good News!!!   I had sent a mail to artifactory team and explained the issue - it seems like they have made a tweak to the HA config that has perhaps resolved the problem!  

I am still testing but first things look good - I have already asked them for the specifics of what they changed so that I can report back here
